File: Question 0212 - Word Search II/src/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution
{
    private void dfs(char[][] board, int r, int c, TrieNode node, ArrayList<String> ans, StringBuilder word, int[][] dir, boolean[][] visited)
    {
        if (r < 0 || c < 0 || r >= board.length || c >= board[r].length || visited[r][c]) return;

        if (node.children[board[r][c]- 'a'] == null) return;

        char ch = board[r][c];
        visited[r][c] = true;
        word.append(ch);

        node = node.children[ch-'a'];
        if (node.isTerminal)
        {
            ans.add(word.toString());
            node.isTerminal = false;
        }

        for(int[] d : dir)
        {
            int newX = r + d[0];
            int newY = c + d[1];
            dfs(board, newX, newY, node, ans, word, dir, visited);
        }

        word.deleteCharAt(word.length()-1);
        visited[r][c] = false;

    }
    public List<String> findWords(char[][] board, String[] words)
    {
        Trie t = new Trie();
        for (String word : words)
        {
            t.insertWord(word);
        }

        ArrayList<String> ans = new ArrayList<>();
        int[][] dir = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        boolean[][] visited = new boolean[board.length][board[0].length];

        for (int i = 0; i < board.length; i++)
        {
            for (int j = 0; j < board[i].length; j++)
            {
                StringBuilder word = new StringBuilder();
                dfs(board, i, j, t.root , ans, word, dir, visited);
            }
        }

        return ans;
    }
}

File: Question 0212 - Word Search II/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class TrieNode
{
public:
    char data;
    TrieNode *children[26];
    bool isTerminal;

    TrieNode(char data)
    {
        this->data = data;

        for (int i = 0; i < 26; i++)
        {
            children[i] = NULL;
        }

        isTerminal = false;
    }
};

class Trie
{
private:
    TrieNode *root;

    void addUtil(TrieNode *root, string word)
    {
        if (word.empty())
        {
            root->isTerminal = true;
            return;
        }

        int idx = word[0] - 'a';
        if (root->children[idx] == NULL)
            root->children[idx] = new TrieNode(word[0]);

        addUtil(root->children[idx], word.substr(1));
    }

public:
    Trie()
    {
        root = new TrieNode('\0');
    }

    void add(string word)
    {
        addUtil(root, word);
    }

    TrieNode *getRoot()
    {
        return root;
    }
};

class Solution
{
private:
    void solve(vector<vector<char>> &board, vector<vector<bool>> &vis, string currWord, set<string> &ans, TrieNode *t, int r, int c, vector<vector<int>> &dir)
    {
        if (r < 0 || r >= board.size() || c < 0 || c >= board[0].size() || vis[r][c])
            return;

        char ch = board[r][c];
        int idx = ch - 'a';

        if (!t->children[idx])
            return;

        vis[r][c] = true;
        currWord.push_back(ch);
        t = t->children[idx];

        if (t->isTerminal)
        {
            ans.insert(currWord);
        }

        for (auto &d : dir)
        {
            int x = r + d[0];
            int y = c + d[1];

            solve(board, vis, currWord, ans, t, x, y, dir);
        }

        vis[r][c] = false;
        currWord.pop_back();
    }

public:
    vector<string> findWords(vector<vector<char>> &board, vector<string> &words)
    {
        Trie *trie = new Trie();

        for (const string &w : words)
        {
            trie->add(w);
        }

        TrieNode *t = trie->getRoot();
        int n = board.size(), m = board[0].size();
        set<string> ans;
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        vector<vector<int>> dir = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                string currWord = "";
                solve(board, vis, currWord, ans, t, i, j, dir);
            }
        }

        vector<string> res(ans.begin(), ans.end());
        return res;
    }
};